File: backend/burn_model.py
'''
Dependency-light inference for the "Will my house burn?" Keras model.

The model in sequential-model-2.h5 is a plain Sequential stack
(Dense -> BatchNorm -> Dropout ... -> softmax). Loading it through TensorFlow
pulls in a ~500 MB dependency just to do a few matrix multiplies, so this
module reads the weights with h5py and runs the forward pass in NumPy.

The notebook (californa-juypter/model2.ipynb) trained the network on
StandardScaler-normalised features but never saved the scaler, so the
original API fed raw label indices into a model that expected z-scores.
SCALER_MEAN / SCALER_SCALE below were recovered by re-running the notebook's
exact split (train_test_split(test_size=0.2, random_state=42)) on
cleaned_dataset.csv; see export_model.py.
'''
import json
import os

import numpy as np

HERE = os.path.dirname(os.path.abspath(__file__))
MODEL_PATH = os.path.join(HERE, 'sequential-model-2.h5')
SCALER_PATH = os.path.join(HERE, 'scaler.json')

BN_EPSILON = 1e-3  # Keras BatchNormalization default


def _load_layers(path=MODEL_PATH):
    import h5py  # local import so the module can be imported without h5py for tooling

    layers = []
    with h5py.File(path, 'r') as f:
        config = json.loads(f.attrs['model_config'])
        weights = f['model_weights']
        for layer in config['config']['layers']:
            kind, cfg = layer['class_name'], layer['config']
            name = cfg['name']
            if kind == 'Dense':
                g = weights[name][name]
                layers.append(('dense', np.array(g['kernel:0']), np.array(g['bias:0']), cfg['activation']))
            elif kind == 'BatchNormalization':
                g = weights[name][name]
                layers.append(('bn', np.array(g['gamma:0']), np.array(g['beta:0']),
                               np.array(g['moving_mean:0']), np.array(g['moving_variance:0'])))
            # InputLayer and Dropout are no-ops at inference time
    return layers


def _load_scaler(path=SCALER_PATH):
    with open(path, 'r', encoding='utf-8') as f:
        data = json.load(f)
    return np.array(data['mean']), np.array(data['scale'])


class BurnModel:
    '''Numpy re-implementation of the trained Keras model.'''

    def __init__(self, model_path=MODEL_PATH, scaler_path=SCALER_PATH):
        self.layers = _load_layers(model_path)
        self.mean, self.scale = _load_scaler(scaler_path)

    def predict(self, x, scaled=False):
        x = np.asarray(x, dtype=np.float64)
        if not scaled:
            x = (x - self.mean) / self.scale
        for layer in self.layers:
            if layer[0] == 'dense':
                _, kernel, bias, activation = layer
                x = x @ kernel + bias
                if activation == 'relu':
                    x = np.maximum(x, 0)
                elif activation == 'softmax':
                    x = np.exp(x - x.max(axis=-1, keepdims=True))
                    x = x / x.sum(axis=-1, keepdims=True)
            else:
                _, gamma, beta, mean, var = layer
                x = gamma * (x - mean) / np.sqrt(var + BN_EPSILON) + beta
        return x

File: Frontend/src/lib/burnModel.js
// Runs the trained Keras model (exported by backend/export_model.py) in the browser.
let modelPromise;

function loadModel() {
  if (!modelPromise) {
    modelPromise = fetch('/model/burn-model.json').then((r) => {
      if (!r.ok) throw new Error('model not found');
      return r.json();
    }).catch((e) => {
      modelPromise = null; // allow a retry
      throw e;
    });
  }
  return modelPromise;
}

export function preloadModel() {
  loadModel().catch(() => {});
}

function forward(model, input) {
  let x = input.map((v, i) => (v - model.scaler.mean[i]) / model.scaler.scale[i]);
  for (const layer of model.layers) {
    if (layer.type === 'dense') {
      const out = layer.bias.slice();
      for (let i = 0; i < x.length; i++) {
        const row = layer.kernel[i];
        const xi = x[i];
        for (let j = 0; j < out.length; j++) out[j] += xi * row[j];
      }
      if (layer.activation === 'relu') x = out.map((v) => Math.max(0, v));
      else if (layer.activation === 'softmax') {
        const m = Math.max(...out);
        const e = out.map((v) => Math.exp(v - m));
        const s = e.reduce((a, b) => a + b, 0);
        x = e.map((v) => v / s);
      } else x = out;
    } else {
      x = x.map((v, i) => layer.gamma[i] * (v - layer.mean[i]) / Math.sqrt(layer.variance[i] + layer.epsilon) + layer.beta[i]);
    }
  }
  return x;
}

export async function predictLocally(form) {
  const model = await loadModel();
  const input = model.features.map((name) => {
    if (name === 'age') return Number(form.age) || 0;
    const entry = Object.entries(model.mappings[name]).find(([, label]) => label === form[name]);
    if (!entry) throw new Error(`Unknown value for ${name}`);
    return Number(entry[0]);
  });
  const probs = forward(model, input);
  const labels = Object.keys(model.mappings.damage).sort((a, b) => a - b).map((k) => model.mappings.damage[k]);
  const best = probs.indexOf(Math.max(...probs));
  return { predicted_damage: labels[best], probabilities: [probs], labels };
}
